### DISCUSSION TITLE
Updating lockfiles

### DISCUSSION THREAD

**Discussion by PurpleBooth** [gembaadvantage/uplift#96] [Q&A]
I was seeing about migrating one of my cargo projects to uplift and I came across an interesting problem. This project is written in rust, and changing the version number in `Cargo.toml` requires the `Cargo.lock`. Normally this is handled by running `cargo fetch` or `cargo update --workspace`.

Is there any way to include this in the version bump commit?

**Comment by ga-paul-t** [comment]
So in this instance, `uplift` needs to be able to execute a bespoke command as part of the `bump` process? I have been toying with the idea recently of introducing hooks into `uplift` that will allow the invocation of shell commands. Please let me know if I have understood this correctly or not. Don't have much experience of using rust

**Reply by PurpleBooth** [reply]
Spot on! That's how `versio`, a similar tool, handles it too.

**Reply by ga-paul-t** [reply]
Great. I'll take a look

**Reply by ga-paul-t** [reply]
Wow, the time has really flown. Didn't realise how old this thread was. Lots of work has been going on to make uplift a more viable product for CI solutions. Just thought I would drop a message that support for hooks should be arriving imminently. If you are still using uplift of course 😄

**Reply by ga-paul-t** [reply]
Dropped the first iteration of hook support. Plans to expand it shortly. Interested in hearing any feedback

**Reply by PurpleBooth** [reply]
Now got this working on on a NodeJS repo and it works great! Thanks! Very smooth very easy

**Reply by ga-paul-t** [reply]
If we go back to your original discussion. In the scenario with `rust` and `cargo`, are having hooks before (and soon after) uplift executing suitable? Or do you need the commands to run straight after the file has been bumped? I have never used `rust` before so have a lack of knowledge in that domain

**Reply by PurpleBooth** [reply]
It worked perfectly, @tomphp and I gave it a go over on [specdown](https://github.com/specdown/specdown), just as easy as with JavaScript, having said that I did miss having a "toml" thing, like the "json" key bump

**Reply by ga-paul-t** [reply]
I agree. I want to add a YAML and TOML equivalent to file bumping. Make that whole process much easier. Leaving regex for more niche use cases

**Reply by PurpleBooth** [reply]
The regex count 0 thing was very handy, replaced multiple instances of the version in a markdown file

**Reply by PurpleBooth** [reply]
Just noticed that the hooks seem to run still when doing a dry run, in our case this makes the repository dirty stopping it from running right

**Reply by ga-paul-t** [reply]
Right. Will raise a bug and sort that out later today

**Reply by ga-paul-t** [reply]
Been thinking about this some more. I guess for the time being I will just switch off the execution of the hooks in `dry-run` mode. But it feels like I need to do something more involved in this area. Ideally, I'd want the hooks to run in `dry-run` mode to give a better experience of what happens during the release process. I guess in the scenario with `cargo` it is changing a tracked file before uplift runs, hence the failed git check. Might be able to refine the git checks further

**Reply by tomphp** [reply]
My feeling at the moment is that `before` is the wrong place to do the type of actions we are talking about (i.e. it's essentially part of the bumping). For this particular use, `before` should be run just before/after the `bump` (after the `gitcheck` task). Maybe a better approach is to add another specific hook (`on_bump`/`before_bump`/`after_bump`/`before_commit`) which can be used for this purpose.

**Reply by ga-paul-t** [reply]
This does sound like the ultimate way to go. A much richer hook lifecycle throughout the entire workflow of uplift. That would provide it with a greater degree of flexibility.

**Reply by ga-paul-t** [reply]
Captured this as an issue, https://github.com/gembaadvantage/uplift/issues/192

**Reply by ga-paul-t** [reply]
Hooks are now skipped in dry run mode

**Reply by ga-paul-t** [reply]
The number of lifecycle hooks has now increased within uplift. You can now invoke shell commands and scripts at many different points within the release workflow.
